### PR TITLE
netdev_ieee802154_submac: unify legacy and radio HAL bootstrap

### PR DESCRIPTION
**Pull request by jia200x**
### Contribution description
This PR unifies the init procedure for the Radio HAL and legacy `netdev_<device>` interface. This allocates a `netdev_ieee802154_submac_t` descriptor if module `netdev_ieee802154_submac` is present. If it's not present, it allocates the legacy `netdev_ieee802154_t` only if the Radio HAL is not present. This ensures that no `netdev` dependencies are pulled if only the Radio HAL is used (e.g OpenWSN).

This whole step allows to run all radios that implement the Radio HAL with `netdev_ieee802154_submac` without changing the bootstrap code. For instance, #15132 should work with this PR.

TL;DR: `USEMODULE+=netdev_ieee802154_submac` should work out of the box for all radios that implement the Radio HAL (e.g LWIP, OpenThread can bootstrap a radio in the same way they used to do with `netdev`)

### Testing procedure
- Murdock should do its work compiling `tests/ieee802154_hal` and `tests/ieee802154_submac`.
- Try testing `gnrc_networking` with and without `netdev_ieee802154_submac`.
- Try LWIP with `nrf802154` (the `cc2538_rf` is not yet in the bootstrap list :/)
- Test that addresses don't change if `netdev_ieee802154_submac` is selected.

### Issues/PRs references
#15132

### CODE REVIEW

File: cpu/cc2538/radio/cc2538_rf.c [hunk 1]
```diff
@@ -193,6 +193,11 @@ bool cc2538_on(void)
 
 void cc2538_setup(cc2538_rf_t *dev)
 {
+#if IS_USED(MODULE_IEEE802154_RADIO_HAL)
+    extern ieee802154_dev_t cc2538_rf_dev;
+    netdev_ieee802154_submac_init(&dev->netdev, &cc2538_rf_dev);
+    cc2538_init();
+#else
```
Comment (benpicco) [inline cpu/cc2538/radio/cc2538_rf.c:200]: You could still do the `netdev_register()` in both cases - as long as we have `netdev_ieee802154_submac`

Comment (jia200x) [reply cpu/cc2538/radio/cc2538_rf.c:200]: Yes, true. I will try that

**Comment by jia200x** [comment]
done! it works now with `netdev_register`

**Comment by benpicco** [comment]
Please squash.

File: sys/net/link_layer/ieee802154/submac.c [hunk 2]
```diff
@@ -300,8 +301,8 @@ int ieee802154_submac_init(ieee802154_submac_t *submac)
     ieee802154_radio_request_on(dev);
 
     /* generate EUI-64 and short address */
-    luid_get_eui64(&submac->ext_addr);
-    luid_get_short(&submac->short_addr);
```
Comment (benpicco) [inline sys/net/link_layer/ieee802154/submac.c:304]: You can also drop the `luid` dependency now.

Comment (jia200x) [reply sys/net/link_layer/ieee802154/submac.c:304]: true!

File: cpu/nrf52/radio/nrf802154/nrf802154_radio.c [hunk 3]
```diff
@@ -714,6 +714,17 @@ static int _set_csma_params(ieee802154_dev_t *dev, const ieee802154_csma_be_t *b
     return 0;
 }
 
+void nrf802154_setup(nrf802154_t *dev)
+{
+    (void) dev;
+#if IS_USED(MODULE_NETDEV_IEEE802154_SUBMAC)
+    netdev_t *netdev = (netdev_t*) dev;
+    netdev_register(netdev, NETDEV_NRF802154, 0);
+    netdev_ieee802154_submac_init(&dev->netdev, &nrf802154_hal_dev);
```
Comment (benpicco) [inline cpu/nrf52/radio/nrf802154/nrf802154_radio.c:723]: You'll need to call `netdev_ieee802154_setup()` after register to configure the MAC address.

**Comment by jia200x** [comment]
done! I also dropped the `luid` dependency and squashed directly

File: sys/net/link_layer/ieee802154/submac.c (hunk 2 above)
Comment (jia200x) [reply sys/net/link_layer/ieee802154/submac.c:304]: done!

File: cpu/nrf52/radio/nrf802154/nrf802154.c [hunk 4]
```diff
@@ -535,6 +514,25 @@ void isr_radio(void)
...
+    nrf802154_dev = netdev_ieee802154;
+
+    netdev->driver = &nrf802154_netdev_driver;
+#ifdef MODULE_GNRC
+#ifdef MODULE_GNRC_SIXLOWPAN
+    netdev_ieee802154->proto = GNRC_NETTYPE_SIXLOWPAN;
+#else
+    netdev_ieee802154->proto = GNRC_NETTYPE_UNDEF;
+#endif
+#endif
```
Comment (benpicco) [inline cpu/nrf52/radio/nrf802154/nrf802154.c:530]: Why does the driver have to deal with this?

Comment (jia200x) [reply cpu/nrf52/radio/nrf802154/nrf802154.c:530]: It shouldn't actually. I just copied what was already there. But this should be indeed handled by the upper layer.

Comment (benpicco) [reply cpu/nrf52/radio/nrf802154/nrf802154.c:530]: `netdev/ieee802154.c` already sets those values, so I think we can just drop this here.
Or would you rather do that in a separate PR?

Comment (jia200x) [reply cpu/nrf52/radio/nrf802154/nrf802154.c:530]: I can do it here :)

Comment (jia200x) [reply cpu/nrf52/radio/nrf802154/nrf802154.c:530]: done!

Comment (benpicco) [reply cpu/nrf52/radio/nrf802154/nrf802154.c]: Please squash :wink:

**Review by benpicco — APPROVED** [review]

**Comment by jia200x** [comment]
shall I squash?

**Review by benpicco — CHANGES_REQUESTED** [review]
`nrf52840dk` now gets stuck in a boot loop with `examples/gnrc_networking`.

Works fine when using `netdev_ieee802154_submac`.

File: cpu/nrf52/radio/nrf802154/nrf802154.c [hunk 5]
```diff
@@ -197,7 +178,7 @@ static void _set_chan(uint16_t chan)
     /* Channel map between 2400 MHZ ... 2500 MHz
```
Comment (benpicco) [inline cpu/nrf52/radio/nrf802154/nrf802154.c:178]: The above assertion is triggered, channel = 0

**Comment by jia200x** [comment]
fixed. I squashes and amended directly (it fails because I removed the initialization). Also, nobody was calling `netdev_ieee802154_reset` there o.O

**Comment by benpicco** [comment]
Yup, that fixed it.
btw: when squashing please don't do a `git rebase master`, instead only do a `git rebase HEAD~n` where n is the number of commits you want to go back, or `git rebase <hash of commit before the PR>` so when clicking the [force-pushed](https://github.com/RIOT-OS/RIOT/compare/00b01861157422b4aa578f30740678d4422a5fc4..ca7240e52974c608360ce4d34490a00f0b847862) link only the changes from this PR show up.

I'll give it a try on the `openmote-b` with `cc2538_rf` just to be sure :wink:

**Review by benpicco — APPROVED** [review]
`cc2538_rf` is also still working, both with `netdev_ieee802154_submac` and without.

**Comment by jia200x** [comment]
> btw: when squashing please don't do a git rebase master, instead only do a git rebase HEAD~n where n is the number of commits you want to go back, or git rebase <hash of commit before the PR> so when clicking the force-pushed link only the changes from this PR show up.

Sure!

**Comment by benpicco** [comment]
Strange observation (possibly unrelated):

With `netdev_ieee802154_submac` on `nrf52840dk` I don't get a public address from the border router in `tests/riotboot_flashwrite` - but without it I always do.

 `examples/gnrc_networking` works

**Comment by jia200x** [comment]
hmmm strange. I will check what's going on there

**Comment by jia200x** [comment]
@benpicco did https://github.com/RIOT-OS/RIOT/pull/15237 by chance fix the issue?

**Comment by benpicco** [comment]
Unfortunately not, the issue is still present on current `master`.
